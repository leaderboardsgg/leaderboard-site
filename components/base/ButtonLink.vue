<template>
  <NuxtLink
    :class="classNames"
    :to="linkTo"
    class="button-link"
  >
    <slot />
  </NuxtLink>
</template>

<script>
export default {
  props: {
    classNames: {
      default: '',
      type: String
    },
    linkTo: {
      default: '#',
      type: String
    }
  }
};
</script>

<style>
  @layer components {
    .button-link {
      @apply rounded px-3 py-2 text-current hover:bg-current;
    }
  }
</style>
