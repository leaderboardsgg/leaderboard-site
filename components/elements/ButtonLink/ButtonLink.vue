<template>
  <NuxtLink class="button-link" v-bind="$attrs" v-on="$listeners">
    <slot> Placeholder </slot>
  </NuxtLink>
</template>

<style>
.button-link {
  @apply rounded px-3 py-2 text-current hover:bg-current;
}
</style>
