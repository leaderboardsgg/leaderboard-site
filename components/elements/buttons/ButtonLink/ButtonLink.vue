<script setup lang="ts">
// blargh
</script>

<template>
  <NuxtLink class="button-link" v-bind="$attrs">
    <slot>Placeholder Link Text</slot>
  </NuxtLink>
</template>

<style lang="postcss" scoped>
.button-link {
  @apply rounded px-3 py-2 text-current hover:bg-current;
}
</style>
