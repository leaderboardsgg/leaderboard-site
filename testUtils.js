import Vue from 'vue';
import { RouterLinkStub } from '@vue/test-utils';
import userEvent from '@testing-library/user-event';
import { render } from '@testing-library/vue';

export const NuxtLinkStub = Vue.component('NuxtLinkStub', {
  extends: RouterLinkStub,
  render(h) {
    const temp = h(
      this.tag || 'a',
      {
        attrs: {
          ...this.$vnode.data.attrs,
          ...this.$attrs,
          href: this.to,
        },
      },
      this.$slots.default,
    );

    return temp;
  },
});

export const stubbedRender = (Component, options) => {
  return render(Component, {
    stubs: { NuxtLink: NuxtLinkStub, ...options?.stubs },
    ...options,
  });
};

export * from '@testing-library/vue';
export * from '@vue/test-utils';
export { userEvent as fireEvent };
