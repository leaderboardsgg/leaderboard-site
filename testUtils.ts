import { createElementBlock, Component, ComponentOptions } from 'vue'
import { MountingOptions } from '@vue/test-utils'
import userEvent from '@testing-library/user-event'
import { render, RenderOptions } from '@testing-library/vue'

export const stubbedRender = (
  Component: Component | ComponentOptions,
  options?: RenderOptions | MountingOptions<any>,
) => {
  return render(Component, {
    global: {
      stubs: {
        NuxtLink: createElementBlock('a'),
      },
    },
    ...options,
  })
}

export * from '@testing-library/vue'
export * from '@nuxt/test-utils'
export * from '@vue/test-utils'
export { userEvent as fireEvent }
