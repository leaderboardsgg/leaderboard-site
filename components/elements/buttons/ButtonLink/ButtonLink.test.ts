// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-expect-error
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { setup, $fetch } from '@nuxt/test-utils-edge'

import { describe, test as it } from 'vitest'
import ButtonLink from './ButtonLink.vue'
import { stubbedRender } from '@/testUtils'

describe('<ButtonLink />', async () => {
  await setup({})

  const defaultAttrs = { class: 'custom-link' }
  const defaultProps = { to: 'https://www.test.com/' }

  it('should render without crashing', () => {
    const { unmount } = stubbedRender(ButtonLink, {
      attrs: defaultAttrs,
      props: defaultProps,
      slots: { default: 'Button' },
    })

    unmount()
  })

  it('renders with the correct <slot />', () => {
    const { getByText } = stubbedRender(ButtonLink, {
      attrs: defaultAttrs,
      props: defaultProps,
      slots: { default: 'Any%' },
    })
    expect(getByText('Any%')).toBeInTheDocument()
  })

  it('renders with the passed link', () => {
    const { getByText } = stubbedRender(ButtonLink, {
      attrs: defaultAttrs,
      props: defaultProps,
    })

    const link = getByText('Placeholder Link Text')
    expect(link.tagName).toEqual('A')
    expect(link.getAttribute('to')).toEqual('https://www.test.com/')
  })

  it('renders with the custom classnames', () => {
    stubbedRender(ButtonLink, {
      attrs: {
        ...defaultAttrs,
        class: 'go-fast',
      },
      props: defaultProps,
    })

    expect(document.querySelector('.go-fast')).toBeInTheDocument()
  })
})
