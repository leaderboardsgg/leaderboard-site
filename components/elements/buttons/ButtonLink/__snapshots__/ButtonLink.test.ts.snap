// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonLink /> renders correctly 1`] = `
<a
  class="custom-link"
  href=""
  to="https://www.test.com/"
>
  Button
</a>
`;
